### PR TITLE
Refactor to allow different `DistanceCalculations` to be used in Highlighter classes

### PR DESCRIPTION
**Pull request by tchambers3**
Rather than setting a flag to indicate whether `horizontalDistance` should be used in `closestSelectionDetailByPixel()`, this refactor allows us to pass a type of distance calculation (`euclideanDistance`, `verticalDistance`, or  `horizontalDistance`) when initializing a highlighter. 

s/o @jasdev for pairing on this.

### CODE REVIEW

**Review by jbrennan — APPROVED** [review]

File: Source/Charts/Utils/DistanceCalculation.swift [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+//  Charts
+//
+//  Created by Travis Chambers on 5/15/19.
+//
+
+import Foundation
+
+public typealias DistanceCalculation = (_ x1: CGFloat, _ y1: CGFloat, _ x2: CGFloat, _ y2: CGFloat) -> CGFloat
+
+public let eucledianDistance: DistanceCalculation = { x1, y1, x2, y2 in
```
Comment (jasdev) [inline Source/Charts/Utils/DistanceCalculation.swift]: Realized we misspelled this, should be `euclideanDistance` 😅

File: Source/Charts/Highlight/ChartHighlighter.swift [hunk 2]
```diff
@@ -16,10 +16,13 @@ open class ChartHighlighter : NSObject, IHighlighter
 {
     /// instance of the data-provider
     @objc open weak var chart: ChartDataProvider?
+
+    let distanceCalculation: DistanceCalculation
     
-    @objc public init(chart: ChartDataProvider)
```
Comment (jasdev) [inline Source/Charts/Highlight/ChartHighlighter.swift:20]: Minor, might be helpful to tack some docs on `distanceCalculation` to let peeps know to look at `DistanceCalculation.swift` for common calculation methods.

File: Source/Charts/Highlight/ChartHighlighter.swift [hunk 3]
```diff
@@ -136,7 +138,7 @@ open class ChartHighlighter : NSObject, IHighlighter
         {
             if axis == nil || high.axis == axis
             {
-                let cDistance = usingXDistance ? getDistance(x1: x, y1: y, x2: high.xPx, y2: high.yPx) : getXDistance(x1: x, x2: high.xPx)
+                let cDistance = distanceCalculation(x, y, high.xPx, high.yPx)
```
Comment (jasdev) [inline Source/Charts/Highlight/ChartHighlighter.swift:142]: This is the crux of the refactor, for other folks reading. The calculation method is now injected through `ChartHighlighter.distanceCalculation` versus sending a boolean through to the method (which could get unwieldy down the line, if we introduce more).

**Review by jasdev — APPROVED** [review]

File: Source/Charts/Utils/DistanceCalculation.swift (hunk 1 above)
Comment (tchambers3) [reply Source/Charts/Utils/DistanceCalculation.swift]: Had a feeling we'd misspell something along the way...

**Review by jbrennan — APPROVED** [review]

File: Source/Charts/Highlight/ChartHighlighter.swift (hunk 2 above)
Comment (jbrennan) [reply Source/Charts/Highlight/ChartHighlighter.swift:20]: Would it make sense for the calc methods to be an enum?

Comment (tchambers3) [reply Source/Charts/Highlight/ChartHighlighter.swift:20]:
```diff
diff --git a/Source/Charts/Charts/BarChartView.swift b/Source/Charts/Charts/BarChartView.swift
index ba7a89b..5fdc3cd 100644
--- a/Source/Charts/Charts/BarChartView.swift
+++ b/Source/Charts/Charts/BarChartView.swift
@@ -27,7 +27,7 @@ open class BarChartView: BarLineChartViewBase, BarChartDataProvider
         
         renderer = BarChartRenderer(dataProvider: self, animator: _animator, viewPortHandler: _viewPortHandler)
         
-        self.highlighter = BarHighlighter(chart: self, distanceCalculation: horizontalDistance)
+        self.highlighter = BarHighlighter(chart: self, distanceCalculation: .horizontal)
         
         self.xAxis.spaceMin = 0.5
         self.xAxis.spaceMax = 0.5
diff --git a/Source/Charts/Charts/HorizontalBarChartView.swift b/Source/Charts/Charts/HorizontalBarChartView.swift
index 6ccf183..57fa9b2 100644
--- a/Source/Charts/Charts/HorizontalBarChartView.swift
+++ b/Source/Charts/Charts/HorizontalBarChartView.swift
@@ -31,7 +31,7 @@ open class HorizontalBarChartView: BarChartView
         rightYAxisRenderer = YAxisRendererHorizontalBarChart(viewPortHandler: _viewPortHandler, yAxis: rightAxis, transformer: _rightAxisTransformer)
         xAxisRenderer = XAxisRendererHorizontalBarChart(viewPortHandler: _viewPortHandler, xAxis: _xAxis, transformer: _leftAxisTransformer, chart: self)
         
-        self.highlighter = HorizontalBarHighlighter(chart: self, distanceCalculation: horizontalDistance)
+        self.highlighter = HorizontalBarHighlighter(chart: self, distanceCalculation: .horizontal)
     }
     
     internal override func calculateLegendOffsets(offsetLeft: inout CGFloat, offsetTop: inout CGFloat, offsetRight: inout CGFloat, offsetBottom: inout CGFloat)
diff --git a/Source/Charts/Highlight/ChartHighlighter.swift b/Source/Charts/Highlight/ChartHighlighter.swift
index d641b7c..cf7b12a 100644
--- a/Source/Charts/Highlight/ChartHighlighter.swift
+++ b/Source/Charts/Highlight/ChartHighlighter.swift
@@ -20,7 +20,7 @@ open class ChartHighlighter : NSObject, IHighlighter
     /// For different `distanceCalculation` methods refer to `DistanceCalculation.swift`
     let distanceCalculation: DistanceCalculation
     
-    @objc public init(chart: ChartDataProvider, distanceCalculation: @escaping DistanceCalculation = euclideanDistance)
+    @objc public init(chart: ChartDataProvider, distanceCalculation: DistanceCalculation = .euclidean)
     {
         self.chart = chart
         self.distanceCalculation = distanceCalculation
@@ -139,7 +139,7 @@ open class ChartHighlighter : NSObject, IHighlighter
         {
             if axis == nil || high.axis == axis
             {
-                let cDistance = distanceCalculation(x, y, high.xPx, high.yPx)
+                let cDistance = distanceCalculation.calculate(x, y, high.xPx, high.yPx)
 
                 if cDistance < distance
                 {
diff --git a/Source/Charts/Highlight/CombinedHighlighter.swift b/Source/Charts/Highlight/CombinedHighlighter.swift
index ab4aeca..b402d8d 100644
--- a/Source/Charts/Highlight/CombinedHighlighter.swift
+++ b/Source/Charts/Highlight/CombinedHighlighter.swift
@@ -23,7 +23,7 @@ open class CombinedHighlighter: ChartHighlighter
         super.init(chart: chart)
         
         // if there is BarData, create a BarHighlighter
-        self.barHighlighter = barDataProvider.barData == nil ? nil : BarHighlighter(chart: barDataProvider, distanceCalculation: horizontalDistance)
+        self.barHighlighter = barDataProvider.barData == nil ? nil : BarHighlighter(chart: barDataProvider, distanceCalculation: .horizontal)
     }
     
     open override func getHighlights(xValue: Double, x: CGFloat, y: CGFloat) -> [Highlight]
diff --git a/Source/Charts/Utils/DistanceCalculation.swift b/Source/Charts/Utils/DistanceCalculation.swift
index 23a9786..d8e5f35 100644
--- a/Source/Charts/Utils/DistanceCalculation.swift
+++ b/Source/Charts/Utils/DistanceCalculation.swift
@@ -7,19 +7,37 @@
 
 import Foundation
 
-public typealias DistanceCalculation = (_ x1: CGFloat, _ y1: CGFloat, _ x2: CGFloat, _ y2: CGFloat) -> CGFloat
+internal typealias DistanceEquation = (_ x1: CGFloat, _ y1: CGFloat, _ x2: CGFloat, _ y2: CGFloat) -> CGFloat
 
 /// Straight line distance between two points
-public let euclideanDistance: DistanceCalculation = { x1, y1, x2, y2 in
+internal let euclideanDistance: DistanceEquation = { x1, y1, x2, y2 in
     hypot(x1 - x2, y1 - y2)
 }
 
 /// Horizontal distance between two points
-public let horizontalDistance: DistanceCalculation = { x1 , _, x2, _ in
+internal let horizontalDistance: DistanceEquation = { x1 , _, x2, _ in
     abs(x1 - x2)
 }
 
 /// Vertical distance between two points
-public let verticalDistance: DistanceCalculation = { _, y1, _, y2 in
+internal let verticalDistance: DistanceEquation = { _, y1, _, y2 in
     abs(y1 - y2)
 }
+
+@objc public enum DistanceCalculation: Int {
+	case euclidean = 0
+	case horizontal
+	case vertical
+
+	var calculate: DistanceEquation {
+		switch self {
+		case .euclidean:
+			return euclideanDistance
+		case .horizontal:
+			return horizontalDistance
+		case .vertical:
+			return verticalDistance
+		}
+	}
+
+}
```

Comment (tchambers3) [reply Source/Charts/Highlight/ChartHighlighter.swift:20]: @jasdev @jbrennan we could do something like this

Comment (jasdev) [reply Source/Charts/Highlight/ChartHighlighter.swift:20]: Niiice!  think using the enumeration simply as a namespace (with static members) would be more flexible in allowing consumers of the library (us, hah) to pass in any calculation method without requiring a PR against `Charts`. Maybe like this?

```patch
diff --git a/Charts.xcodeproj/project.pbxproj b/Charts.xcodeproj/project.pbxproj
index f9a2897..8420998 100644
--- a/Charts.xcodeproj/project.pbxproj
+++ b/Charts.xcodeproj/project.pbxproj
@@ -24,7 +24,7 @@
 		1311BEC21E9CC264E971EFAF /* ILineRadarChartDataSet.swift in Sources */ = {isa = PBXBuildFile; fileRef = 4F7E6A99D82E6AE3804D5A39 /* ILineRadarChartDataSet.swift */; };
 		135F11CE20425AF600D655A3 /* PieChartTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 135F11CD20425AF600D655A3 /* PieChartTests.swift */; };
 		146EE16342C2BADC92E45BF2 /* ILineScatterCandleRadarChartDataSet.swift in Sources */ = {isa = PBXBuildFile; fileRef = 9249AD9AEC8C85772365A128 /* ILineScatterCandleRadarChartDataSet.swift */; };
-		15325869228C6B5B003FB915 /* DistanceCalculation.swift in Sources */ = {isa = PBXBuildFile; fileRef = 15325868228C6B5A003FB915 /* DistanceCalculation.swift */; };
+		15325869228C6B5B003FB915 /* Distance.swift in Sources */ = {isa = PBXBuildFile; fileRef = 15325868228C6B5A003FB915 /* Distance.swift */; };
 		17E994DA88777AA1D8CCFC58 /* BarChartDataSet.swift in Sources */ = {isa = PBXBuildFile; fileRef = C31AA65EA27776F8C653C7E8 /* BarChartDataSet.swift */; };
 		203A39685CC96FC625F616E4 /* IHighlighter.swift in Sources */ = {isa = PBXBuildFile; fileRef = 998F2BFE318471AFC05B50AC /* IHighlighter.swift */; };
 		219192CA6B4895319AB49DCA /* BarLineScatterCandleBubbleRenderer.swift in Sources */ = {isa = PBXBuildFile; fileRef = 5B1C588E9DF6FFD56D7ADF8E /* BarLineScatterCandleBubbleRenderer.swift */; };
@@ -181,7 +181,7 @@
 		11EF1FE22549E885C8F40738 /* CombinedChartView.swift */ = {isa = PBXFileReference; includeInIndex = 1; lastKnownFileType = sourcecode.swift; name = CombinedChartView.swift; path = Source/Charts/Charts/CombinedChartView.swift; sourceTree = "<group>"; };
 		12409C3EA15787C11AF0D2BC /* ChartDataEntryBase.swift */ = {isa = PBXFileReference; includeInIndex = 1; lastKnownFileType = sourcecode.swift; name = ChartDataEntryBase.swift; path = Source/Charts/Data/Implementations/Standard/ChartDataEntryBase.swift; sourceTree = "<group>"; };
 		135F11CD20425AF600D655A3 /* PieChartTests.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; name = PieChartTests.swift; path = Tests/Charts/PieChartTests.swift; sourceTree = "<group>"; };
-		15325868228C6B5A003FB915 /* DistanceCalculation.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; name = DistanceCalculation.swift; path = Source/Charts/Utils/DistanceCalculation.swift; sourceTree = "<group>"; };
+		15325868228C6B5A003FB915 /* Distance.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; name = Distance.swift; path = Source/Charts/Utils/Distance.swift; sourceTree = "<group>"; };
 		18462BFDD9DEE76D51D40503 /* ScatterChartView.swift */ = {isa = PBXFileReference; includeInIndex = 1; lastKnownFileType = sourcecode.swift; name = ScatterChartView.swift; path = Source/Charts/Charts/ScatterChartView.swift; sourceTree = "<group>"; };
 		18BFB0A14A5C47A302A597D9 /* ICandleChartDataSet.swift */ = {isa = PBXFileReference; includeInIndex = 1; lastKnownFileType = sourcecode.swift; name = ICandleChartDataSet.swift; path = Source/Charts/Data/Interfaces/ICandleChartDataSet.swift; sourceTree = "<group>"; };
 		1C02C3AF5C92FCFC18224C35 /* XAxisRenderer.swift */ = {isa = PBXFileReference; includeInIndex = 1; lastKnownFileType = sourcecode.swift; name = XAxisRenderer.swift; path = Source/Charts/Renderers/XAxisRenderer.swift; sourceTree = "<group>"; };
@@ -589,7 +589,7 @@
 				FF475B9593B9898853814340 /* Transformer.swift */,
 				324C9127B53A8D39C8B49277 /* TransformerHorizontalBarChart.swift */,
 				72EAEBB7CF73E33565FC2896 /* ViewPortHandler.swift */,
-				15325868228C6B5A003FB915 /* DistanceCalculation.swift */,
+				15325868228C6B5A003FB915 /* Distance.swift */,
 			);
 			name = Utils;
 			sourceTree = "<group>";
@@ -932,7 +932,7 @@
 				796D3E63A37A95FD9D1AB9A1 /* ChevronDownShapeRenderer.swift in Sources */,
 				F103D90FC5DEEA0D7BB4407E /* ChevronUpShapeRenderer.swift in Sources */,
 				B85DEB06B4C1AFFC8A0E3295 /* CircleShapeRenderer.swift in Sources */,
-				15325869228C6B5B003FB915 /* DistanceCalculation.swift in Sources */,
+				15325869228C6B5B003FB915 /* Distance.swift in Sources */,
 				0529DD51622C8769C1121F90 /* CrossShapeRenderer.swift in Sources */,
 				8A9FF54E2075A9047CC8E953 /* IShapeRenderer.swift in Sources */,
 				8EF7B3FBE37F72CC030CD865 /* SquareShapeRenderer.swift in Sources */,
diff --git a/Source/Charts/Charts/BarChartView.swift b/Source/Charts/Charts/BarChartView.swift
index ba7a89b..8648c20 100644
--- a/Source/Charts/Charts/BarChartView.swift
+++ b/Source/Charts/Charts/BarChartView.swift
@@ -27,7 +27,7 @@ open class BarChartView: BarLineChartViewBase, BarChartDataProvider
         
         renderer = BarChartRenderer(dataProvider: self, animator: _animator, viewPortHandler: _viewPortHandler)
         
-        self.highlighter = BarHighlighter(chart: self, distanceCalculation: horizontalDistance)
+        self.highlighter = BarHighlighter(chart: self, distanceCalculation: Distance.horizontal)
         
         self.xAxis.spaceMin = 0.5
         self.xAxis.spaceMax = 0.5
diff --git a/Source/Charts/Charts/HorizontalBarChartView.swift b/Source/Charts/Charts/HorizontalBarChartView.swift
index 6ccf183..5167785 100644
--- a/Source/Charts/Charts/HorizontalBarChartView.swift
+++ b/Source/Charts/Charts/HorizontalBarChartView.swift
@@ -31,7 +31,7 @@ open class HorizontalBarChartView: BarChartView
         rightYAxisRenderer = YAxisRendererHorizontalBarChart(viewPortHandler: _viewPortHandler, yAxis: rightAxis, transformer: _rightAxisTransformer)
         xAxisRenderer = XAxisRendererHorizontalBarChart(viewPortHandler: _viewPortHandler, xAxis: _xAxis, transformer: _leftAxisTransformer, chart: self)
         
-        self.highlighter = HorizontalBarHighlighter(chart: self, distanceCalculation: horizontalDistance)
+        self.highlighter = HorizontalBarHighlighter(chart: self, distanceCalculation: Distance.horizontal)
     }
     
     internal override func calculateLegendOffsets(offsetLeft: inout CGFloat, offsetTop: inout CGFloat, offsetRight: inout CGFloat, offsetBottom: inout CGFloat)
diff --git a/Source/Charts/Highlight/ChartHighlighter.swift b/Source/Charts/Highlight/ChartHighlighter.swift
index d641b7c..022b003 100644
--- a/Source/Charts/Highlight/ChartHighlighter.swift
+++ b/Source/Charts/Highlight/ChartHighlighter.swift
@@ -18,9 +18,9 @@ open class ChartHighlighter : NSObject, IHighlighter
     @objc open weak var chart: ChartDataProvider?
 
     /// For different `distanceCalculation` methods refer to `DistanceCalculation.swift`
-    let distanceCalculation: DistanceCalculation
+    let distanceCalculation: Distance.Calculation
     
-    @objc public init(chart: ChartDataProvider, distanceCalculation: @escaping DistanceCalculation = euclideanDistance)
+    @objc public init(chart: ChartDataProvider, distanceCalculation: @escaping Distance.Calculation = Distance.euclidean)
     {
         self.chart = chart
         self.distanceCalculation = distanceCalculation
diff --git a/Source/Charts/Highlight/CombinedHighlighter.swift b/Source/Charts/Highlight/CombinedHighlighter.swift
index ab4aeca..6e6abac 100644
--- a/Source/Charts/Highlight/CombinedHighlighter.swift
+++ b/Source/Charts/Highlight/CombinedHighlighter.swift
@@ -23,7 +23,7 @@ open class CombinedHighlighter: ChartHighlighter
         super.init(chart: chart)
         
         // if there is BarData, create a BarHighlighter
-        self.barHighlighter = barDataProvider.barData == nil ? nil : BarHighlighter(chart: barDataProvider, distanceCalculation: horizontalDistance)
+        self.barHighlighter = barDataProvider.barData == nil ? nil : BarHighlighter(chart: barDataProvider, distanceCalculation: Distance.horizontal)
     }
     
     open override func getHighlights(xValue: Double, x: CGFloat, y: CGFloat) -> [Highlight]
diff --git a/Source/Charts/Utils/Distance.swift b/Source/Charts/Utils/Distance.swift
new file mode 100644
index 0000000..b569048
--- /dev/null
+++ b/Source/Charts/Utils/Distance.swift
@@ -0,0 +1,32 @@
+//
+//  Distance.swift
+//  Charts
+//
+//  Created by Travis Chambers on 5/15/19.
+//
+
+import Foundation
+
+public enum Distance {
+    public typealias Calculation = (
+        _ x1: CGFloat,
+        _ y1: CGFloat,
+        _ x2: CGFloat,
+        _ y2: CGFloat
+    ) -> CGFloat
+
+    /// Straight line distance between two points.
+    public static let euclidean: Calculation = { x1, y1, x2, y2 in
+        hypot(x1 - x2, y1 - y2)
+    }
+
+    /// Horizontal distance between two points.
+    public static let horizontal: Calculation = { x1 , _, x2, _ in
+        abs(x1 - x2)
+    }
+
+    /// Vertical distance between two points.
+    public static let vertical: Calculation = { _, y1, _, y2 in
+        abs(y1 - y2)
+    }
+}
diff --git a/Source/Charts/Utils/DistanceCalculation.swift b/Source/Charts/Utils/DistanceCalculation.swift
deleted file mode 100644
index 23a9786..0000000
--- a/Source/Charts/Utils/DistanceCalculation.swift
+++ /dev/null
@@ -1,25 +0,0 @@
-//
-//  DistanceCalculation.swift
-//  Charts
-//
-//  Created by Travis Chambers on 5/15/19.
-//
-
-import Foundation
-
-public typealias DistanceCalculation = (_ x1: CGFloat, _ y1: CGFloat, _ x2: CGFloat, _ y2: CGFloat) -> CGFloat
-
-/// Straight line distance between two points
-public let euclideanDistance: DistanceCalculation = { x1, y1, x2, y2 in
-    hypot(x1 - x2, y1 - y2)
-}
-
-/// Horizontal distance between two points
-public let horizontalDistance: DistanceCalculation = { x1 , _, x2, _ in
-    abs(x1 - x2)
-}
-
-/// Vertical distance between two points
-public let verticalDistance: DistanceCalculation = { _, y1, _, y2 in
-    abs(y1 - y2)
-}

```

**Review by jbrennan — APPROVED** [review]
Love how this turned out

**Pull request merged**